### PR TITLE
Implement external approval with topics

### PR DESCRIPTION
**Pull request by mvo-dev**
# Changes
## Implement external approval
It is now possible to create JIT constraints based on approval from non-peer groups. This is achieved through two new conditions
`has({}.externalApprovalConstraint)` and `has({}.reviewerPrivilege)`. Users/groups with reviewer privilege can review the activation requests from users with an external approval constraint (provided that it is for the same role).
## Implement topic
It is now possible to create fine grained peers and external reviewers by adding a 'topic' to the constraint. Only peers/reviewers of the same topic can review the activation request. Topics are specified after the constraint for instance `has({}.multiPartyApprovalConstraint.cvation_team)` specifies the topic `cvation_team`. If no topic is listed then access is granted on all topics.
## Diagram
image

### CODE REVIEW

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java [hunk 1]
```diff
@@ -35,615 +36,727 @@
...
-      .findEntitlements(
-        eq(SAMPLE_REQUESTING_USER),
-        eq(SAMPLE_PROJECT),
-        eq(EnumSet.of(ActivationType.MPA)),
-        eq(EnumSet.of(Entitlement.Status.AVAILABLE))))
-      .thenReturn(new EntitlementSet<>(
-        new TreeSet<>(Set.of(mpaEntitlement)),
-        Set.of(),
-        Set.of()));
+        .findRequesterPrivileges(
```
Comment (johannesbertoft) [inline sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java:322]: I'm a little confused by this test, maybe you can explain the idea here.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java [hunk 2]
```diff
@@ -35,615 +36,727 @@
...
+    assertDoesNotThrow(
+        () -> catalog.verifyUserCanActivateRequesterPrivileges(
+            SAMPLE_REQUESTING_USER,
+            SAMPLE_PROJECT,
+            peerApproval,
+            List.of(peerApprovalPrivilege.id())));
+  }
+
+  @Test
+  public void whenRequestHasTopicEmpty_ThenVerifyUserCanActivateRequesterPrivilegesThrows()
```
Comment (johannesbertoft) [inline sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java:338]: Something is confusing with the naming here, because the previous test is called the same thing but ends in not throwing, but this test ends in throwing, but the "when" sounds to be exactly the same. let's discuss

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestProjectRoleActivator.java [hunk 3]
```diff
@@ -58,32 +64,37 @@ public class TestProjectRoleActivator {
...
 
-    verify(resourceManagerClient, times(2))
-      .addProjectIamBinding(
-        eq(SAMPLE_PROJECT),
-        argThat(b ->
-            IamTemporaryAccessConditions.isTemporaryAccessCondition(b.getCondition().getExpression()) &&
-            b.getCondition().getTitle().equals(JitConstraints.ACTIVATION_CONDITION_TITLE)),
-        eq(EnumSet.of(ResourceManagerClient.IamBindingOptions.PURGE_EXISTING_TEMPORARY_BINDINGS)),
-        eq("Self-approved, justification: justification"));
+    verify(resourceManagerClient, times(1))
```
Comment (johannesbertoft) [inline sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestProjectRoleActivator.java:90]: Why this change from 2 to 1? What does it mean in this context?

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationRequest.java [hunk 4]
```diff
@@ -27,49 +27,58 @@
 import java.time.Duration;
 import java.time.Instant;
 import java.util.Collection;
-import java.util.Set;
-import java.util.stream.Collectors;
 
 /**
- * Represents a request for activating one or more entitlements.
+ * Represents a request for activating one or more requester privileges.
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationRequest.java:32]: You are correct.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/TestActivationId.java [hunk 5]
```diff
@@ -39,8 +39,8 @@ public void toStringReturnsId() {
 
   @Test
   public void toStringContainsTypePrefix() {
-    var id = ActivationId.newId(ActivationType.MPA);
-    assertTrue(id.toString().startsWith("mpa-"));
+    var id = ActivationId.newId(new PeerApproval("test"));
+    assertTrue(id.toString().startsWith("peer_approval(test)-"));
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/TestActivationId.java:43]: The id is combined as `activation_type-random_id`.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationType.java [hunk 6]
```diff
@@ -21,13 +21,8 @@
 
 package com.google.solutions.jitaccess.core.catalog;
 
-public enum ActivationType {
-  /** Entitlement can be activated using self-approval */
-  JIT,
+public abstract class ActivationType {
+  public abstract String name();
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationType.java:25]: Makes sense.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    if (!(other instanceof ExternalApproval)) {
+      return false;
+    }
+
+    if (this.topic.equals("")) {
+      return true;
+    }
+
+    return this.name().equals(other.name());
+  }
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java:28]: Will update the naming here.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestAssetInventoryRepository.java [hunk 8]
```diff
@@ -169,409 +188,586 @@ public void whenEffectiveIamPoliciesContainsInapplicableBindings_ThenFindProject
...
-  // findEntitlements.
-  //---------------------------------------------------------------------------
+  // ---------------------------------------------------------------------------
+  // findRequesterPrivileges.
+  // ---------------------------------------------------------------------------
 
   @Test
-  public void whenEffectiveIamPoliciesContainEligibleBindings_ThenFindEntitlementsReturnsList() throws Exception {
+  public void whenEffectiveIamPoliciesContainEligibleBindings_ThenFindRequesterPrivilegesReturnsList()
+      throws Exception {
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestAssetInventoryRepository.java:299]: Yeah, would be better to split up.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java (hunk 1 above)
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java:322]: This test tests that a user with a `RequesterPrivilege` of `PeerApproval("")` can activate a RequesterPrivilege of type `PeerApproval("topic").

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java (hunk 2 above)
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java:338]: One is named `whenUserHasTopicEmpty` whereas the other is `whenRequestHasTopicEmpty`. This test tests that a user with `RequesterPrivilege` of type `PeerApproval("topic")` **cannot** activate a RequesterPrivilege of type `PeerApproval("")`.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestProjectRoleActivator.java (hunk 3 above)
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestProjectRoleActivator.java:90]: Before we were testing that when we added IAM bindings for an `ActivationRequest` with two associated roles, then two IAM bindings were added. Since now we can only have one associated role also only one IAM binding should be added.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java (hunk 2 above)
Comment (johannesbertoft) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestMpaProjectRoleCatalog.java:338]: Understood

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/RequesterPrivilegeSet.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Set of requester privileges
+ *
+ * @param availableRequesterPrivileges available requester privileges,
+ *                                     regardless of status
+ * @param activeRequesterPrivilegeIds  IDs of active requester privileges.
+ * @param warnings                     encountered warnings, if any.
+ */
+public record RequesterPrivilegeSet<TId extends PrivilegeId>(
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/RequesterPrivilegeSet.java:39]: Will fix.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.regex.Pattern;
+
+/**
+ * Helper class for creating privileges and parsing IAM conditions.
+ */
+class PrivilegeFactory {
+  /** Condition title for activated role bindings */
+  public static final String ACTIVATION_CONDITION_TITLE = "JIT access activation";
+
+  public static final String VALID_TOPIC_PATTERN = "(.([a-zA-Z][a-zA-Z0-9-_]*))?";
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:43]: Will think about updating the regex.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  /**
+   * Condition that marks a role binding as eligible for self approver privilege
+   */
+  private static final Pattern SELF_APPROVER_CONDITION_PATTERN = Pattern
+      .compile("^\\s*has\\(\\s*\\{\\s*\\}.jitaccessconstraint\\s*\\)\\s*$");
+
+  /** Condition that marks a role binding as eligible for peer privilege */
+  private static final Pattern PEER_CONDITION_PATTERN = Pattern
+      .compile("^\\s*has\\(\\s*\\{\\s*\\}.multipartyapprovalconstraint" + VALID_TOPIC_PATTERN
+          + "\\s*\\)\\s*$");
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:54]: Yeah does not make sense, so only for backwards compatibility.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+      return false;
+    }
+
+    // Strip all whitespace to simplify expression matching.
+    var expression = iamCondition
+        .getExpression()
+        .toLowerCase()
+        .replace(" ", "");
+
+    return pattern.matcher(expression).matches();
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:79]: No, we can remove that.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+  /** Check if the IAM condition is an external approval constraint */
+  private static boolean isRequesterCondition(Expr iamCondition) {
+    return isMatchingCondition(iamCondition, REQUESTER_CONDITION_PATTERN);
+  }
+
+  /** Check if the IAM condition is a reviewer privilege */
+  private static boolean isReviewerCondition(Expr iamCondition) {
+    return isMatchingCondition(iamCondition, REVIEWER_CONDITION_PATTERN);
+  }
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:116]: Will update to clean up.

File: sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java [hunk 14]
```diff
@@ -569,151 +614,141 @@ public ActivationStatusResponse requestActivation(
...
-      obfuscatedActivationToken != null && !obfuscatedActivationToken.trim().isEmpty(),
-      "An activation token is required");
+        obfuscatedActivationToken != null && !obfuscatedActivationToken.trim().isEmpty(),
+        "An activation token is required");
 
     var activationToken = TokenObfuscator.decode(obfuscatedActivationToken);
     var iapPrincipal = (UserPrincipal) securityContext.getUserPrincipal();
 
-    MpaActivationRequest<ProjectRoleBinding> activationRequest;
+    com.google.solutions.jitaccess.core.catalog.ActivationRequest<ProjectRoleBinding> activationRequest;
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java:677]: There is a local class named `ActivationRequest` which shadows the real class. Will update naming.

File: sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java [hunk 15]
```diff
@@ -995,33 +1003,31 @@ public String getType() {
...
-        ((MpaActivationRequest<ProjectRoleBinding>)activation.request()).reviewers(), // Move reviewers to CC.
-        String.format(
-          "%s requests access to project %s",
-          activation.request().requestingUser(),
-          projectId));
+          List.of(activation.request().requestingUser()),
+          activation.request().reviewers(), // Move
+                                            // reviewers
+                                            // to
+                                            // CC.
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java:1015]: Will fix.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  public boolean contains(ActivationType other) {
+    if (!(other instanceof ExternalApproval)) {
+      return false;
+    }
+
+    if (this.topic.equals("")) {
+      return true;
+    }
+
+    return this.name().equals(other.name());
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java:27]: Changed field name to `baseName`

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationRequest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationRequest.java:32]: Fixed

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationType.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ActivationType.java:25]: Updated to interface

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/ExternalApproval.java:28]: Changed to `isParentTypeOf`

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/RequesterPrivilegeSet.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/RequesterPrivilegeSet.java:39]: Updated

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:43]: Fixed the regex. Turns out that there was an issue of `.` matching any character rather than `\.` matching only a dot.

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:79]: Removed

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:116]: Removed unnecessary methods

File: sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java [hunk 17]
```diff
@@ -569,151 +614,141 @@ public ActivationStatusResponse requestActivation(
...
 
       return new ActivationStatusResponse(
-        iapPrincipal.getId(),
-        activationRequest,
-        Entitlement.Status.ACTIVATION_PENDING); // TODO(later): Could check if's been activated already.
-    }
-    catch (Exception e) {
+          iapPrincipal.getId(),
+          List.of(activationRequest),
+          RequesterPrivilege.Status.ACTIVATION_PENDING); // TODO(later): Could check if's been activated
```
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java:641]: Fixed

File: sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java:677]: Renamed local class

File: sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/web/rest/ApiResource.java]: Fixed

File: sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/main/java/com/google/solutions/jitaccess/core/catalog/project/PrivilegeFactory.java:79]: Turns out the regex was needed to remove tab and new line characters so I have removed the unnecessary replace call.

File: sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestAssetInventoryRepository.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mvo-dev) [reply sources/src/test/java/com/google/solutions/jitaccess/core/catalog/project/TestAssetInventoryRepository.java:299]: Split up

**Review by johannesbertoft — APPROVED** [review]
Nice refactoring work! Approved
